llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp

```diff
@@ -5453,14 +5453,32 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
     // Multiply and Add Packed Words
     //   < 4 x i32> @llvm.x86.sse2.pmadd.wd(<8 x i16>, <8 x i16>)
     //   < 8 x i32> @llvm.x86.avx2.pmadd.wd(<16 x i16>, <16 x i16>)
-
+    //   <16 x i32> @llvm.x86.avx512.pmaddw.d.512(<32 x i16>, <32 x i16>)
+    //
     // Multiply and Add Packed Signed and Unsigned Bytes
     //   < 8 x i16> @llvm.x86.ssse3.pmadd.ub.sw.128(<16 x i8>, <16 x i8>)
     //   <16 x i16> @llvm.x86.avx2.pmadd.ub.sw(<32 x i8>, <32 x i8>)
+    //   <32 x i16> @llvm.x86.avx512.pmaddubs.w.512(<64 x i8>, <64 x i8>)
+    //
+    // These intrinsics are auto-upgraded into non-masked forms:
```

Comment:
how does one upgrade into a non-masked for? that's overapproximating, right?

Reply:
The auto-upgrade is performed by an earlier LLVM step (https://llvm.org/doxygen/AutoUpgrade_8cpp_source.html). Essentially, LLVM converts those masked pmadd intrinsics into an equivalent combination of non-masked pmadd intrinsics + masking operations. MemorySanitizer never sees the masked intrinsics (in fact, I believe you can't even specify them using with Intrinsic::x86_ anymore). Thus, compared to non-masked pmadd intrinsics, there is no additional loss of precision when instrumenting the auto-upgraded masked pmadd intrinsics.